arch/x86/entry.S

```diff
@@ -32,70 +32,100 @@
 #include <usermode.h>
 #include <errno.h>
 
-.macro _handle_usermode cr3
+.macro MASK_USER_FLAGS
+    PUSHF
+    andl $~(USERMODE_FLAGS_MASK), (%_ASM_SP)
+    POPF
+.endm
+
+.macro _from_usermode switch_stack=1
+    SET_CR3 cr3
+    swapgs
+    .if \switch_stack == 1
+        SWITCH_STACK
+    .endif
+.endm
+
+.macro _to_usermode switch_stack=1
+    .if \switch_stack == 1
+        SWITCH_STACK
+    .endif
+    swapgs
+    SET_CR3 user_cr3
+.endm
+
+.macro exc_from_usermode
     testb $0x3, cpu_exc_cs(%_ASM_SP)
     jz 1f /* skip if from kernel mode */
-        SET_CR3 \cr3
-        swapgs
+        MASK_USER_FLAGS
+        _from_usermode switch_stack=0
     1:
 .endm
 
-.macro enter_from_usermode
-    _handle_usermode cr3
-.endm
-
-.macro enter_to_usermode
-    _handle_usermode user_cr3
+.macro exc_to_usermode
+    testb $0x3, cpu_exc_cs(%_ASM_SP)
+    jz 1f /* skip if from kernel mode */
+        _to_usermode switch_stack=0
+    1:
 .endm
 
 .macro syscall_from_usermode
-    SET_CR3 cr3
-    swapgs
-    SWITCH_STACK
+    _from_usermode
 .endm
 
 .macro syscall_to_usermode
-    SWITCH_STACK
-    swapgs
-    SET_CR3 user_cr3
+    _to_usermode
+.endm
+
+.macro irq_from_usermode
+    testb $0x3, cpu_exc_cs(%_ASM_SP)
+    jz 1f /* skip if from kernel mode */
+        MASK_USER_FLAGS
+        _from_usermode
+        /* Preserve USER CS and fake: RIP, error_code, vector on kernel stack */
+        /* This is expected by the irq_to_usermode to detect context          */
+        push $__USER_CS
+        sub $cpu_exc_cs, %_ASM_SP
+    1:
```

Comment:
The main reason is we do not use IST for anything but #DF. So for exceptions the stack stays the same for both user and kernel. But in this case it is fine, since exception for usermode is supposed to terminate the task. So for simplicity, we piggy back on the user stack, handle the `do_exception()` stuff and terminate it. And we do so, cause CPU put the exception frame on user stack already and we just want to use it.
For IRQs we do switch the stack for user, hence the different macro set. The CPU put the irq frame onto user stack, so we make it similar to the exception frame, switch to kernel (and use the hack to let the `irq_to_usermode()` detect the context in case of the user transition), handle the interrupt and switch back for the IRET.

Reply:
Naaah, Mathias is right. The CPU uses TSS's RSP0 when IST is not specified. Just verified empirically. Going to make the changes much simpler now.